### ISSUE TITLE
Onload with MiMalloc conflict

### ISSUE THREAD

**Issue by CPlusDave** [microsoft/mimalloc#713]
Solarflare (->Xilinix->AMD) Onload hooks the open(...) system call. This causes a problem during initial startup when mimalloc calls...
####
static void os_detect_overcommit(void) {
#if defined(__linux__)
  int fd = open("/proc/sys/vm/overcommit_memory", O_RDONLY);
####
Similar bug/issue in jemalloc fixed here : https://github.com/jemalloc/jemalloc/issues/443

Onload intercepts the open() call, tries to do a malloc when Onload realizes that it is being initialized for the first time, and we get back into mimalloc until it fails at...
####
#2  0x000000000042b7d5 in _mi_assert_fail (assertion=0x122416e "_mi_heap_main.cookie != 0", fname=0x122406b "mimalloc-internal.h", line=442, func=0x122415f "_mi_ptr_cookie") at options.c:411
No locals.
#3  0x0000000000413f21 in _mi_ptr_cookie (p=0x20000000000) at mimalloc-internal.h:442
        __func__ = "_mi_ptr_cookie"
#4  0x000000000041fc8f in mi_segment_alloc (required=0, page_alignment=0, req_arena_id=0, tld=0x172d0c0 <tld_main+32>, os_tld=0x172d450 <tld_main+944>, huge_page=0x0) at segment.c:902
####

Earlier in this stack here is the hook around issue.

#19 0x000000000042c9b1 in mi_heap_calloc (heap=0x172dfe0 <_mi_heap_main>, count=1, size=32) at alloc.c:647
        total = 32
#20 0x000000000042829a in mi_calloc (count=1, size=32) at alloc.c:651
No locals.
#21 0x00007f7aefe92a15 in _dlerror_run () from /lib64/libdl.so.2
No symbol table info available.
#22 0x00007f7aefe923c3 in dlsym () from /lib64/libdl.so.2
No symbol table info available.
#23 0x00007f7af08d8b57 in citp_syscall_init () from /lib64/libonload.so
No symbol table info available.
#24 0x00007f7af08d8713 in citp_do_init () from /lib64/libonload.so
No symbol table info available.
#25 0x00007f7af08decc2 in open64 () from /lib64/libonload.so
No symbol table info available.
#26 0x00000000004179a6 in os_detect_overcommit () at os.c:257
        fd = 32766
        buf = "\000\000\000\000\000\000\000\000\000\205\374Ұ\023\324\b\020\377f\230\376\177\000\000\270\327\307\357z\177\000"
        nread = 140731455307552
#27 0x0000000000417a47 in _mi_os_init () at os.c:286

read() and close() are also intercepted and will have the same issues.

**Comment by CPlusDave** [comment]
tldr; here is the jemalloc fix

```
diff --git a/src/pages.c b/src/pages.c
index 2a9b7e3..a33e7d7 100644
--- a/src/pages.c
+++ b/src/pages.c
@@ -214,11 +214,14 @@ os_overcommits_proc(void)
        char buf[1];
        ssize_t nread;

-       fd = open("/proc/sys/vm/overcommit_memory", O_RDONLY);
+       fd = syscall(SYS_open, "/proc/sys/vm/overcommit_memory", O_RDONLY);
        if (fd == -1)
                return (false); /* Error. */

-       nread = read(fd, &buf, sizeof(buf));
+       nread = syscall(SYS_read, fd, &buf, sizeof(buf));
+
+       syscall(SYS_close, fd);
+
        if (nread < 1)
                return (false); /* Error. */
        /*
```

**Comment by rigtorp** [comment]
Are you linking mimalloc statically or dynamically?

**Comment by CPlusDave** [comment]
mimalloc static, object-file linking in my example. Onload DLL is always done via LD_PRELOAD. Don't think it will matter much how mimalloc is linked, as the LD_PRELOAD on Onload will intercept the open(), read(), close() calls.

**Comment by rigtorp** [comment]
I have the same problem in a different context and since I was using only
single NUMA node cloud instances I patched mimalloc to return static values
for vm.overcommit and number of NUMA nodes.
